ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/trielog/TrieLogPruner.java

```diff
@@ -0,0 +1,188 @@
+    final Stream<byte[]> trieLogKeys = rootWorldStateStorage.streamTrieLogKeys(loadingLimit);
+    final AtomicLong count = new AtomicLong();
+    trieLogKeys.forEach(
+        blockHashAsBytes -> {
+          Hash blockHash = Hash.wrap(Bytes32.wrap(blockHashAsBytes));
+          final Optional<BlockHeader> header = blockchain.getBlockHeader(blockHash);
+          if (header.isPresent()) {
+            trieLogBlocksAndForksByDescendingBlockNumber.put(header.get().getNumber(), blockHash);
+            count.getAndIncrement();
+          } else {
+            // prune orphaned blocks (sometimes created during block production)
+            rootWorldStateStorage.pruneTrieLog(blockHash);
+          }
+        });
+    LOG.atInfo().log("Loaded {} trie logs from database", count);
+    pruneFromCache();
+  }
+
+  void cacheForLaterPruning(final long blockNumber, final Hash blockHash) {
+    LOG.atTrace()
+        .setMessage("caching trie log for later pruning blockNumber {}; blockHash {}")
+        .addArgument(blockNumber)
+        .addArgument(blockHash)
+        .log();
+    trieLogBlocksAndForksByDescendingBlockNumber.put(blockNumber, blockHash);
+  }
+
+  void pruneFromCache() {
+
+    final long retainAboveThisBlock = blockchain.getChainHeadBlockNumber() - numBlocksToRetain;
+    final long retainAboveThisBlockOrFinalized =
+        blockchain
+            .getFinalized()
+            .flatMap(blockchain::getBlockHeader)
+            .map(ProcessableBlockHeader::getNumber)
+            .map(finalizedBlock -> Math.min(finalizedBlock, retainAboveThisBlock))
+            .orElse(retainAboveThisBlock);
+
+    LOG.atTrace()
+        .setMessage(
+            "min((chainHeadNumber: {} - numBlocksToRetain: {}) = {}, finalized: {})) = retainAboveThisBlockOrFinalized: {}")
+        .addArgument(blockchain::getChainHeadBlockNumber)
+        .addArgument(numBlocksToRetain)
+        .addArgument(retainAboveThisBlock)
+        .addArgument(
+            () ->
+                blockchain
+                    .getFinalized()
+                    .flatMap(blockchain::getBlockHeader)
+                    .map(ProcessableBlockHeader::getNumber)
+                    .orElse(null))
+        .addArgument(retainAboveThisBlockOrFinalized)
+        .log();
+
+    final var pruneWindowEntries =
+        trieLogBlocksAndForksByDescendingBlockNumber.asMap().entrySet().stream()
+            .dropWhile((e) -> e.getKey() > retainAboveThisBlockOrFinalized)
+            .limit(pruningLimit);
+
+    final List<Long> blockNumbersToRemove = new ArrayList<>();
+
+    final AtomicInteger count = new AtomicInteger();
+    pruneWindowEntries.forEach(
+        (e) -> {
+          for (Hash blockHash : e.getValue()) {
+            rootWorldStateStorage.pruneTrieLog(blockHash);
+            count.getAndIncrement();
+          }
+          blockNumbersToRemove.add(e.getKey());
+        });
+
+    blockNumbersToRemove.forEach(trieLogBlocksAndForksByDescendingBlockNumber::removeAll);
```

Comment:
If there is an error with pruning a trielog could end up with an inconsistency in the trieLogBlocksAndForksByDescendingBlockNumber map since only some trielogs might be removed.

Reply:
Good catch, I think if StorageException occurs, we will bubble up too far. Catching this error inside the foreach should allow me to keep `blockNumbersToRemove` consistent. Any trie logs that missed pruning due to error will remain in the cache and will be picked up in the next pruner execution (next block).

Reply:
Handled this here https://github.com/hyperledger/besu/pull/6026/commits/4681a858386dc9165b9850b427a1a26bfbbbc500